isb_lib/opencontext_adapter.py

```diff
@@ -0,0 +1,175 @@
+import datetime
+import time
+
+import isb_lib.core
+import logging
+import requests
+import igsn_lib.models.thing
+import typing
+import dateparser
+
+HTTP_TIMEOUT = 10.0  # seconds
+OPENCONTEXT_API = "https://opencontext.org/subjects-search/.json?add-attribute-uris=1&attributes=obo-foodon-00001303%2Coc-zoo-has-anat-id%2Ccidoc-crm-p2-has-type%2Ccidoc-crm-p45-consists-of%2Ccidoc-crm-p49i-is-former-or-current-keeper-of%2Ccidoc-crm-p55-has-current-location%2Cdc-terms-temporal%2Cdc-terms-creator%2Cdc-terms-contributor&prop=oc-gen-cat-sample-col%7C%7Coc-gen-cat-bio-subj-ecofact%7C%7Coc-gen-cat-object&response=metadata%2Curi-meta&sort=updated--desc"
+MEDIA_JSON = "application/json"
+
+
+def get_logger():
+    return logging.getLogger("isb_lib.opencontext_adapter")
+
+
+class OpenContextItem(object):
+    AUTHORITY_ID = "OPENCONTEXT"
+
+    def __init__(self, identifier: str, source):
+        self.identifier = identifier
+        self.item = source
+
+    def as_thing(
+        self,
+        t_created: datetime.datetime,
+        status: int,
+        resolved_url: str,
+        t_resolved: datetime.datetime,
+        resolve_elapsed: float,
+        media_type: str = None,
+    ) -> igsn_lib.models.thing.Thing:
+        L = get_logger()
+        L.debug("OpenContextItem.asThing")
+        # Note: SESAR incorrectly returns "application/json;charset=UTF-8" for json-ld content
+        if media_type is None:
+            media_type = MEDIA_JSON
+        _thing = igsn_lib.models.thing.Thing(
+            id=self.identifier,
+            tcreated=t_created,
+            item_type=None,
+            authority_id=OpenContextItem.AUTHORITY_ID,
+            resolved_url=resolved_url,
+            resolved_status=status,
+            tresolved=t_resolved,
+            resolve_elapsed=resolve_elapsed,
+        )
+        if not isinstance(self.item, dict):
+            L.error("Item is not an object")
+            return _thing
+        _thing.item_type = "sample"
+        _thing.related = None
+        _thing.resolved_media_type = media_type
+        # _thing.resolve_elapsed = resolve_elapsed
```

Comment:
Right now the way this is hooked up, I can't really wire this in.  If it's important it's certainly possible, but it looked like it was convenient for SESAR because we had a secondary http request to get details for a particular record.  Since we get them all at once in opencontext, we don't have access to the raw http response at the call site.